Add Winters et al matrix dissipation for 1D, 2D Euler

jlchan opened the pull request:
Adding the stationary contact-preserving matrix dissipation operator of Winters, Derigs, Gassner, and Walch (2017) from https://doi.org/10.1016/j.jcp.2016.12.006. Only `CompressibleEulerEquations1D`, `CompressibleEulerEquations2D` are implemented at the moment. 

The 1D and 2D implementations are directly adapted from [Atum.jl](https://github.com/mwarusz/Atum.jl/blob/c7ed44f2b7972ac726ef345da7b98b0bda60e2a3/src/balancelaws/euler.jl#L198). The 3D implementation is adapted from [FLUXO](https://github.com/project-fluxo/fluxo) by @andrewwinters5000

ranocha commented on src/equations/numerical_fluxes.jl:
```diff
@@ -261,6 +261,23 @@ function Base.show(io::IO, d::DissipationLaxFriedrichsEntropyVariables)
+@doc raw"""
+    MatrixDissipationWintersEtal()
+
+Creates the Roe-like entropy stable matrix dissipation operator from Winters et al (2017). This operator
```
```suggestion
Creates the Roe-like entropy stable matrix dissipation operator from Winters et al. (2017). This operator
```

ranocha commented on src/equations/numerical_fluxes.jl:
````diff
@@ -261,6 +261,23 @@ function Base.show(io::IO, d::DissipationLaxFriedrichsEntropyVariables)
+    MatrixDissipationWintersEtal()
+
+Creates the Roe-like entropy stable matrix dissipation operator from Winters et al (2017). This operator
+must be used together with an entropy-conservative two-point flux function (e.g., `flux_ec`) to yield 
+an entropy-stable surface flux. The surface flux function can be initialized as:
+```julia
+flux_es = FluxPlusDissipation(flux_ec, MatrixDissipationWintersEtal())
+```
````
````suggestion
must be used together with an entropy-conservative two-point flux function 
(e.g., [`flux_ranocha`](@ref) or [`flux_chandrashekar`](@ref)) to yield 
an entropy-stable surface flux. The surface flux function can be initialized as:
```julia
flux_es = FluxPlusDissipation(flux_ec, MatrixDissipationWintersEtal())
```
````

`flux_ec` is not defined for the Euler equations, is it?

ranocha commented on src/equations/numerical_fluxes.jl:
```diff
@@ -261,6 +261,23 @@ function Base.show(io::IO, d::DissipationLaxFriedrichsEntropyVariables)
+  for high Mach number ideal MHD and compressible Euler simulations (2024). Journal of Computational Physics.
+  [DOI: 10.1016/j.jcp.2016.12.006](https://doi.org/10.1016/j.jcp.2016.12.006).
+"""
+struct MatrixDissipationWintersEtal end
```
Consistency with the other dissipation operators would suggest to use something like `DissipationMatrixWintersEtal`. @sloede?

sloede replied:
Right. Unless "matrix dissipation" is really the "noun" we want here, because there are multiple "matrix dissipation" types. If in doubt, I'd go with @ranocha's suggestion, but I'll leave it up to you

jlchan replied:
Thanks, that makes sense. I started with `DissipationWintersEtal` but thought that might be too general - will update the PR.

andrewwinters5000 reviewed (changes requested):
Thanks for adding this @jlchan ! My only additional thought would be to add the 3D version for the Euler equations while we are at it.

jlchan commented on src/equations/numerical_fluxes.jl:
````diff
@@ -261,6 +261,25 @@ function Base.show(io::IO, d::DissipationLaxFriedrichsEntropyVariables)
+```
+This implementation is adapted from the [Atum.jl library](https://github.com/mwarusz/Atum.jl/blob/c7ed44f2b7972ac726ef345da7b98b0bda60e2a3/src/balancelaws/euler.jl#L198).
+For the derivation of the matrix dissipation operator, see:
+- A. Winters, D. Derigs, G. Gassner, S. Walch, A uniquely defined entropy stable matrix dissipation operator 
+  for high Mach number ideal MHD and compressible Euler simulations (2024). Journal of Computational Physics.
````
Oops - thanks for catching that.

jlchan commented on test/test_unit.jl:
```diff
@@ -1637,6 +1637,64 @@ end
+    u_rr = prim2cons(SVector(1.1, -0.2, 0, 2.0), equations)
+    v_ll = cons2entropy(u_ll, equations)
+    v_rr = cons2entropy(u_rr, equations)
+    @test dissipation_matrix_winters_etal(u_ll, u_rr, SVector(1.0, 0.0), equations)[[
+                                                                                        1,
+                                                                                        2,
+                                                                                        4
+                                                                                    ]] ≈
```
Yup, no problem.

andrewwinters5000 commented:
@jlchan I took the liberty of preparing what I think is a working version of the 3D variant. It would need an additional test or so but it is in the file below

<details>
<summary>
<tt>wdgw-matrix-diss.jl.control</tt>
</summary>

```julia
@inline function (dissipation::DissipationMatrixWintersEtal)(u_ll, u_rr,
                                                             normal_direction::AbstractVector,
                                                             equations::CompressibleEulerEquations3D)
    (; gamma) = equations

    norm_ = norm(normal_direction)
    unit_normal_direction = normal_direction / norm_

    rho_ll, v1_ll, v2_ll, v3_ll, p_ll = cons2prim(u_ll, equations)
    rho_rr, v1_rr, v2_rr, v3_rr, p_rr = cons2prim(u_rr, equations)

    b_ll = rho_ll / (2 * p_ll)
    b_rr = rho_rr / (2 * p_rr)

    rho_log = ln_mean(rho_ll, rho_rr)
    b_log = ln_mean(b_ll, b_rr)
    v1_avg = avg(v1_ll, v1_rr)
    v2_avg = avg(v2_ll, v2_rr)
    v3_avg = avg(v3_ll, v3_rr)
    p_avg = avg(rho_ll, rho_rr) / (2 * avg(b_ll, b_rr))
    v_squared_bar = v1_ll * v1_rr + v2_ll * v2_rr + v3_ll * v3_rr
    h_bar = gamma / (2 * b_log * (gamma - 1)) + 0.5f0 * v_squared_bar
    c_bar = sqrt(gamma * p_avg / rho_log)

    v_avg_normal = dot(SVector(v1_avg, v2_avg, v3_avg), unit_normal_direction)

    lambda_1 = abs(v_avg_normal - c_bar) * rho_log / (2 * gamma)
    lambda_2 = abs(v_avg_normal) * rho_log * (gamma - 1) / gamma
    lambda_3 = abs(v_avg_normal + c_bar) * rho_log / (2 * gamma)
    lambda_4 = abs(v_avg_normal) * p_avg # scaled repeated eigenvalue in the tangential direction

    v1_minus_c = v1_avg - c_bar * unit_normal_direction[1]
    v2_minus_c = v2_avg - c_bar * unit_normal_direction[2]
    v3_minus_c = v3_avg - c_bar * unit_normal_direction[3]
    v1_plus_c = v1_avg + c_bar * unit_normal_direction[1]
    v2_plus_c = v2_avg + c_bar * unit_normal_direction[2]
    v3_plus_c = v3_avg + c_bar * unit_normal_direction[3]
    v1_tangential = v1_avg - v_avg_normal * unit_normal_direction[1]
    v2_tangential = v2_avg - v_avg_normal * unit_normal_direction[2]
    v3_tangential = v3_avg - v_avg_normal * unit_normal_direction[3]

    entropy_vars_jump = cons2entropy(u_rr, equations) - cons2entropy(u_ll, equations)
    entropy_var_rho_jump, entropy_var_rho_v1_jump,
    entropy_var_rho_v2_jump, entropy_var_rho_v3_jump, entropy_var_rho_e_jump = entropy_vars_jump

    velocity_minus_c_dot_entropy_vars_jump = v1_minus_c * entropy_var_rho_v1_jump +
                                             v2_minus_c * entropy_var_rho_v2_jump +
                                             v3_minus_c * entropy_var_rho_v3_jump
    velocity_plus_c_dot_entropy_vars_jump = v1_plus_c * entropy_var_rho_v1_jump +
                                            v2_plus_c * entropy_var_rho_v2_jump +
                                            v3_plus_c * entropy_var_rho_v3_jump
    velocity_avg_dot_vjump = v1_avg * entropy_var_rho_v1_jump +
                             v2_avg * entropy_var_rho_v2_jump +
                             v3_avg * entropy_var_rho_v3_jump
    w1 = lambda_1 * (entropy_var_rho_jump + velocity_minus_c_dot_entropy_vars_jump +
          (h_bar - c_bar * v_avg_normal) * entropy_var_rho_e_jump)
    w2 = lambda_2 * (entropy_var_rho_jump + velocity_avg_dot_vjump +
          v_squared_bar / 2 * entropy_var_rho_e_jump)
    w3 = lambda_3 * (entropy_var_rho_jump + velocity_plus_c_dot_entropy_vars_jump +
          (h_bar + c_bar * v_avg_normal) * entropy_var_rho_e_jump)

    entropy_var_v_normal_jump = dot(SVector(entropy_var_rho_v1_jump,
                                            entropy_var_rho_v2_jump,
                                            entropy_var_rho_v3_jump),
                                    unit_normal_direction)

    dissipation_rho = w1 + w2 + w3

    dissipation_rho_v1 = (w1 * v1_minus_c +
                          w2 * v1_avg +
                          w3 * v1_plus_c +
                          lambda_4 * (entropy_var_rho_v1_jump -
                           unit_normal_direction[1] * entropy_var_v_normal_jump +
                           entropy_var_rho_e_jump * v1_tangential))

    dissipation_rho_v2 = (w1 * v2_minus_c +
                          w2 * v2_avg +
                          w3 * v2_plus_c +
                          lambda_4 * (entropy_var_rho_v2_jump -
                           unit_normal_direction[2] * entropy_var_v_normal_jump +
                           entropy_var_rho_e_jump * v2_tangential))

    dissipation_rho_v3 = (w1 * v3_minus_c +
                          w2 * v3_avg +
                          w3 * v3_plus_c +
                          lambda_4 * (entropy_var_rho_v3_jump -
                           unit_normal_direction[3] * entropy_var_v_normal_jump +
                           entropy_var_rho_e_jump * v3_tangential))

    v_tangential_dot_entropy_vars_jump = v1_tangential * entropy_var_rho_v1_jump +
                                         v2_tangential * entropy_var_rho_v2_jump +
                                         v3_tangential * entropy_var_rho_v3_jump

    dissipation_rhoe = (w1 * (h_bar - c_bar * v_avg_normal) +
                        w2 * 0.5f0 * v_squared_bar +
                        w3 * (h_bar + c_bar * v_avg_normal) +
                        lambda_4 * (v_tangential_dot_entropy_vars_jump +
                         entropy_var_rho_e_jump *
                         (v1_avg^2 + v2_avg^2 + v3_avg^2 - v_avg_normal^2)))

    return -0.5f0 *
           SVector(dissipation_rho, dissipation_rho_v1, dissipation_rho_v2,
                   dissipation_rho_v2, dissipation_rhoe) * norm_
end
```

jlchan commented:
> @jlchan I took the liberty of preparing what I think is a working version of the 3D variant. It would need an additional test or so but it is in the file below
> wdgw-matrix-diss.jl
> 
> ```julia
> @inline function (dissipation::DissipationMatrixWintersEtal)(u_ll, u_rr,
>                                                              normal_direction::AbstractVector,
>                                                              equations::CompressibleEulerEquations3D)
>     (; gamma) = equations
> 
>     norm_ = norm(normal_direction)
>     unit_normal_direction = normal_direction / norm_
> 
>     rho_ll, v1_ll, v2_ll, v3_ll, p_ll = cons2prim(u_ll, equations)
>     rho_rr, v1_rr, v2_rr, v3_rr, p_rr = cons2prim(u_rr, equations)
> 
>     b_ll = rho_ll / (2 * p_ll)
>     b_rr = rho_rr / (2 * p_rr)
> 
>     rho_log = ln_mean(rho_ll, rho_rr)
>     b_log = ln_mean(b_ll, b_rr)
>     v1_avg = avg(v1_ll, v1_rr)
>     v2_avg = avg(v2_ll, v2_rr)
>     v3_avg = avg(v3_ll, v3_rr)
>     p_avg = avg(rho_ll, rho_rr) / (2 * avg(b_ll, b_rr))
>     v_squared_bar = v1_ll * v1_rr + v2_ll * v2_rr + v3_ll * v3_rr
>     h_bar = gamma / (2 * b_log * (gamma - 1)) + 0.5f0 * v_squared_bar
>     c_bar = sqrt(gamma * p_avg / rho_log)
> 
>     v_avg_normal = dot(SVector(v1_avg, v2_avg, v3_avg), unit_normal_direction)
> 
>     lambda_1 = abs(v_avg_normal - c_bar) * rho_log / (2 * gamma)
>     lambda_2 = abs(v_avg_normal) * rho_log * (gamma - 1) / gamma
>     lambda_3 = abs(v_avg_normal + c_bar) * rho_log / (2 * gamma)
>     lambda_4 = abs(v_avg_normal) * p_avg # scaled repeated eigenvalue in the tangential direction
> 
>     v1_minus_c = v1_avg - c_bar * unit_normal_direction[1]
>     v2_minus_c = v2_avg - c_bar * unit_normal_direction[2]
>     v3_minus_c = v3_avg - c_bar * unit_normal_direction[3]
>     v1_plus_c = v1_avg + c_bar * unit_normal_direction[1]
>     v2_plus_c = v2_avg + c_bar * unit_normal_direction[2]
>     v3_plus_c = v3_avg + c_bar * unit_normal_direction[3]
>     v1_tangential = v1_avg - v_avg_normal * unit_normal_direction[1]
>     v2_tangential = v2_avg - v_avg_normal * unit_normal_direction[2]
>     v3_tangential = v3_avg - v_avg_normal * unit_normal_direction[3]
> 
>     entropy_vars_jump = cons2entropy(u_rr, equations) - cons2entropy(u_ll, equations)
>     entropy_var_rho_jump, entropy_var_rho_v1_jump,
>     entropy_var_rho_v2_jump, entropy_var_rho_v3_jump, entropy_var_rho_e_jump = entropy_vars_jump
> 
>     velocity_minus_c_dot_entropy_vars_jump = v1_minus_c * entropy_var_rho_v1_jump +
>                                              v2_minus_c * entropy_var_rho_v2_jump +
>                                              v3_minus_c * entropy_var_rho_v3_jump
>     velocity_plus_c_dot_entropy_vars_jump = v1_plus_c * entropy_var_rho_v1_jump +
>                                             v2_plus_c * entropy_var_rho_v2_jump +
>                                             v3_plus_c * entropy_var_rho_v3_jump
>     velocity_avg_dot_vjump = v1_avg * entropy_var_rho_v1_jump +
>                              v2_avg * entropy_var_rho_v2_jump +
>                              v3_avg * entropy_var_rho_v3_jump
>     w1 = lambda_1 * (entropy_var_rho_jump + velocity_minus_c_dot_entropy_vars_jump +
>           (h_bar - c_bar * v_avg_normal) * entropy_var_rho_e_jump)
>     w2 = lambda_2 * (entropy_var_rho_jump + velocity_avg_dot_vjump +
>           v_squared_bar / 2 * entropy_var_rho_e_jump)
>     w3 = lambda_3 * (entropy_var_rho_jump + velocity_plus_c_dot_entropy_vars_jump +
>           (h_bar + c_bar * v_avg_normal) * entropy_var_rho_e_jump)
> 
>     entropy_var_v_normal_jump = dot(SVector(entropy_var_rho_v1_jump,
>                                             entropy_var_rho_v2_jump,
>                                             entropy_var_rho_v3_jump),
>                                     unit_normal_direction)
> 
>     dissipation_rho = w1 + w2 + w3
> 
>     dissipation_rho_v1 = (w1 * v1_minus_c +
>                           w2 * v1_avg +
>                           w3 * v1_plus_c +
>                           lambda_4 * (entropy_var_rho_v1_jump -
>                            unit_normal_direction[1] * entropy_var_v_normal_jump +
>                            entropy_var_rho_e_jump * v1_tangential))
> 
>     dissipation_rho_v2 = (w1 * v2_minus_c +
>                           w2 * v2_avg +
>                           w3 * v2_plus_c +
>                           lambda_4 * (entropy_var_rho_v2_jump -
>                            unit_normal_direction[2] * entropy_var_v_normal_jump +
>                            entropy_var_rho_e_jump * v2_tangential))
> 
>     dissipation_rho_v3 = (w1 * v3_minus_c +
>                           w2 * v3_avg +
>                           w3 * v3_plus_c +
>                           lambda_4 * (entropy_var_rho_v3_jump -
>                            unit_normal_direction[3] * entropy_var_v_normal_jump +
>                            entropy_var_rho_e_jump * v3_tangential))
> 
>     v_tangential_dot_entropy_vars_jump = v1_tangential * entropy_var_rho_v1_jump +
>                                          v2_tangential * entropy_var_rho_v2_jump +
>                                          v3_tangential * entropy_var_rho_v3_jump
> 
>     dissipation_rhoe = (w1 * (h_bar - c_bar * v_avg_normal) +
>                         w2 * 0.5f0 * v_squared_bar +
>                         w3 * (h_bar + c_bar * v_avg_normal) +
>                         lambda_4 * (v_tangential_dot_entropy_vars_jump +
>                          entropy_var_rho_e_jump *
>                          (v1_avg^2 + v2_avg^2 + v3_avg^2 - v_avg_normal^2)))
> 
>     return -0.5f0 *
>            SVector(dissipation_rho, dissipation_rho_v1, dissipation_rho_v2,
>                    dissipation_rho_v2, dissipation_rhoe) * norm_
> end
> ```

Awesome, thanks! I'll add it into the PR but it probably will be next week (some travel this weekend).

andrewwinters5000 commented:
> > @jlchan I took the liberty of preparing what I think is a working version of the 3D variant. It would need an additional test or so but it is in the file below
> 
> Awesome, thanks! I'll add it into the PR but it probably will be next week (some travel this weekend).

Apologies, ignore the previous code as there were mistakes in the way the tangential directions were handled. I got confused with the heavily extracted way of computing the matrix dissipation term. Below is an implementation reminiscent of how it is done in FLUXO where we use rotational invariance to rotate the solution, compute the dissipation term and then back-rotate the result.

<details>
<summary>
<tt>diss-matrix-3d.jl</tt>
</summary>

```julia
@inline function (dissipation::DissipationMatrixWintersEtal)(u_ll, u_rr,
                                                             normal_direction::AbstractVector,
                                                             equations::CompressibleEulerEquations3D)
    (; gamma) = equations

    # Step 1:
    # Rotate solution into the appropriate direction

    norm_ = norm(normal_direction)
    # Normalize the vector without using `normalize` since we need to multiply by the `norm_` later
    normal_vector = normal_direction / norm_

    # Some vector that can't be identical to normal_vector (unless normal_vector == 0)
    tangent1 = SVector(normal_direction[2], normal_direction[3], -normal_direction[1])
    # Orthogonal projection
    tangent1 -= dot(normal_vector, tangent1) * normal_vector
    tangent1 = normalize(tangent1)

    # Third orthogonal vector
    tangent2 = normalize(cross(normal_direction, tangent1))

    u_ll_rotated = rotate_to_x(u_ll, normal_vector, tangent1, tangent2, equations)
    u_rr_rotated = rotate_to_x(u_rr, normal_vector, tangent1, tangent2, equations)

    # Step 2:
    # Compute the averages using the rotated variables
    rho_ll, v1_ll, v2_ll, v3_ll, p_ll = cons2prim(u_ll_rotated, equations)
    rho_rr, v1_rr, v2_rr, v3_rr, p_rr = cons2prim(u_rr_rotated, equations)

    b_ll = rho_ll / (2 * p_ll)
    b_rr = rho_rr / (2 * p_rr)

    rho_log = ln_mean(rho_ll, rho_rr)
    b_log = ln_mean(b_ll, b_rr)
    v1_avg = avg(v1_ll, v1_rr)
    v2_avg = avg(v2_ll, v2_rr)
    v3_avg = avg(v3_ll, v3_rr)
    p_avg = avg(rho_ll, rho_rr) / (2 * avg(b_ll, b_rr))
    v_squared_bar = v1_ll * v1_rr + v2_ll * v2_rr + v3_ll * v3_rr
    h_bar = gamma / (2 * b_log * (gamma - 1)) + 0.5f0 * v_squared_bar
    c_bar = sqrt(gamma * p_avg / rho_log)

    # Step 3:
    # Build the dissipation term as given in Appendix A of the paper

    # Get entropy variables jump in the rotated variables
    w_jump = cons2entropy(u_rr_rotated, equations) - cons2entropy(u_ll_rotated, equations)

    # Entries of the diagonal scaling matrix where D = ABS(\Lambda)T
    lambda_1 = abs(v1_avg - c_bar) * rho_log / (2 * gamma)
    lambda_2 = abs(v1_avg) * rho_log * (gamma - 1) / gamma
    lambda_3 = abs(v1_avg) * p_avg # scaled repeated eigenvalue in the tangential direction
    lambda_5 = abs(v1_avg + c_bar) * rho_log / (2 * gamma)
    D = SVector(lambda_1, lambda_2, lambda_3, lambda_3, lambda_5)

    # Entries of the right eigenvector matrix (others have already been precomputed)
    r21 = v1_avg - c_bar
    r25 = v1_avg + c_bar
    r51 = h_bar - v1_avg * c_bar
    r52 = 0.5f0 * v_squared_bar
    r55 = h_bar + v1_avg * c_bar

    # Build R and transpose of R matrices
    R = @SMatrix [[1;; 1;; 0;; 0;; 1];
                  [r21;; v1_avg;; 0;; 0;; r25];
                  [v2_avg;; v2_avg;; 1;; 0;; v2_avg];
                  [v3_avg;; v3_avg;; 0;; 1;; v3_avg];
                  [r51;; r52;; v2_avg;; v3_avg;; r55]]

    RT = @SMatrix [[1;; r21;; v2_avg;; v3_avg;; r51];
                   [1;; v1_avg;; v2_avg;; v3_avg;; r52];
                   [0;; 0;; 1;; 0;; v2_avg];
                   [0;; 0;; 0;; 1;; v3_avg];
                   [1;; r25;; v2_avg;; v3_avg;; r55]]

    # Compute the dissipation term R * D * R^T * [[w]] from right-to-left

    # First comes R^T * [[w]]
    diss = RT * w_jump
    # Next multiply with the eigenvalues and Barth scaling
    for j in 1:5
       diss[j] *= D[j]
    end
    # Finally apply the remaining eigenvector matrix
    diss = R * diss

    # Step 4:
    # Do not forget to backrotate my young padawan and then return with proper normalization scaling
    return -0.5f0 * rotate_from_x(diss, normal_vector, tangent1, tangent2, equations) * norm_
end
```

andrewwinters5000 commented on src/equations/numerical_fluxes.jl:
````diff
@@ -261,6 +261,25 @@ function Base.show(io::IO, d::DissipationLaxFriedrichsEntropyVariables)
+```julia
+flux_es = FluxPlusDissipation(flux_ec, DissipationMatrixWintersEtal())
+```
+This implementation is adapted from the [Atum.jl library](https://github.com/mwarusz/Atum.jl/blob/c7ed44f2b7972ac726ef345da7b98b0bda60e2a3/src/balancelaws/euler.jl#L198).
````
If we decide to include the 3D we might also want to mention that the 1D/2D came from Atum.jl and the 3D was helped by FLUXO (https://github.com/project-fluxo/fluxo/blob/master/src/equation/navierstokes/riemann.f90#L901)

andrewwinters5000 reviewed (commented):
Unsure if we will include the 3D version in this PR but I just left a comment about it for completeness.

jlchan commented:
Done adding the 1D, 2D, 3D versions of the matrix dissipation flux, as well as an elixir testing the modified Sod problem. The matrix flux does what is expected and reduces oscillations over regions that the contact discontinuity passes over compared with local Lax Friedrichs. 

Modified Sod with LxF:
![modified_sod_LxF](https://github.com/user-attachments/assets/b04c800f-278c-4be5-8546-d0297cb8cb6e)

Modified Sod with the matrix dissipation:
![modified_sod_matrix_dissipation](https://github.com/user-attachments/assets/4d70f6c0-ce74-467c-a18c-8e7b04da39a8)

ranocha commented on src/equations/numerical_fluxes.jl:
```diff
@@ -261,6 +261,27 @@ function Base.show(io::IO, d::DissipationLaxFriedrichsEntropyVariables)
+  [DOI: 10.1016/j.jcp.2016.12.006](https://doi.org/10.1016/j.jcp.2016.12.006).
+"""
+struct DissipationMatrixWintersEtal end
+@inline avg(u_ll, u_rr) = 0.5f0 * (u_ll + u_rr)
```
I find it confusing to define `avg` here, I have to admit. We use the explicit form everywhere else.

jlchan replied:
Removed

ranocha commented on src/equations/compressible_euler_3d.jl:
```diff
@@ -1237,6 +1237,97 @@ end
+    c_bar = sqrt(gamma * p_avg / rho_log)
+
+    # Step 3:
+    # Build the dissipation term as given in Appendix A of the paper
```
Could you please specify the paper here?

ranocha commented on src/equations/numerical_fluxes.jl:
```diff
@@ -261,6 +261,27 @@ function Base.show(io::IO, d::DissipationLaxFriedrichsEntropyVariables)
+@doc raw"""
+    DissipationMatrixWintersEtal()
+
+Creates the Roe-like entropy stable matrix dissipation operator from Winters et al. (2017). This operator
```
```suggestion
Creates the Roe-like entropy-stable matrix dissipation operator from Winters et al. (2017). This operator
```

as below?

ranocha reviewed (changes requested):
Thanks! Could you please run the formatter?

Did you also test the 2D and 3D versions in full runs, not only in unit tests?

andrewwinters5000 commented on src/equations/compressible_euler_3d.jl:
```diff
@@ -1237,6 +1237,97 @@ end
+
+    rho_log = ln_mean(rho_ll, rho_rr)
+    b_log = ln_mean(b_ll, b_rr)
+    v1_avg = avg(v1_ll, v1_rr)
+    v2_avg = avg(v2_ll, v2_rr)
+    v3_avg = avg(v3_ll, v3_rr)
+    p_avg = avg(rho_ll, rho_rr) / (2 * avg(b_ll, b_rr))
```
```suggestion
    v1_avg = 0.5f0 * (v1_ll + v1_rr)
    v2_avg = 0.5f0 * (v2_ll + v2_rr)
    v3_avg = 0.5f0 * (v3_ll + v3_rr)
    p_avg = 0.5f0 * (rho_ll + rho_rr) / (b_ll + b_rr)
```
Make this more explicit and remove the need for the helper function `avg`

andrewwinters5000 reviewed (changes requested):
Everything looks good, I left one comment to streamline the 3D implementation. It would also be interesting to see a comparison with 2D or 3D Sedov blast wave.

jlchan commented:
> Did you also test the 2D and 3D versions in full runs, not only in unit tests?

Not yet - I'll add those next. Thanks!

jlchan commented:
@andrewwinters5000 I implemented the `orientation::Integer` specialization but noticed some artifacts. 

If I run `trixi_include(joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_blast_wave.jl"), tspan=(0,.50))` with LxF dissipation, I get the following plot:
![lxf_blast](https://github.com/user-attachments/assets/6591d4c6-8cb4-451c-aab2-637762449caa)

If I run `trixi_include(joinpath(examples_dir(), "tree_2d_dgsem", "elixir_euler_blast_wave.jl"), surface_flux=FluxPlusDissipation(flux_ranocha, DissipationMatrixWintersEtal()), cfl=0.5, tspan=(0.0, .50))` with matrix dissipation I get the following instead:
![matrix_dissipation_blast](https://github.com/user-attachments/assets/cb9fabf7-0bf5-4196-bdf4-79bb1a06ffa0)

Do you think this is expected?

ranocha commented on src/equations/numerical_fluxes.jl:
```diff
@@ -280,7 +280,34 @@ For the derivation of the matrix dissipation operator, see:
+                                                             equations::AbstractEquations{2})
+    if orientation == 1
+        return dissipation(u_ll, u_rr, SVector(1, 0), equations)
+    elseif orientation == 2
```
```suggestion
    else # orientation == 2
```

Otherwise, we get a type instability since `nothing` is returned depending on the value of `orientation`.

ranocha commented on src/equations/numerical_fluxes.jl:
```diff
@@ -280,7 +280,34 @@ For the derivation of the matrix dissipation operator, see:
+        return dissipation(u_ll, u_rr, SVector(1, 0, 0), equations)
+    elseif orientation == 2
+        return dissipation(u_ll, u_rr, SVector(0, 1, 0), equations)
+    elseif orientation == 3
```
```suggestion
    else # orientation == 3
```

ranocha commented on test/test_tree_2d_euler.jl:
```diff
@@ -290,6 +290,24 @@ end
                             1.831228461662809
                         ],
                         maxiters=30)
+
+    @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_blast_wave.jl"),
+                        surface_flux=FluxPlusDissipation(flux_ranocha,
+                                                         DissipationMatrixWintersEtal()),
+                        cfl=0.5, tspan=(0.0, 0.01),
+                        l2=[
+                            0.0240344741630761,
+                            0.026530873906038414,
+                            0.026530644332116007,
+                            0.08614440613707544
+                        ],
+                        linf=[
+                            1.9656605818508415,
+                            1.217300641837901,
+                            1.217300641837901,
+                            2.9173277369453516
+                        ])
+
```
```suggestion

    # Ensure that we do not have excessive memory allocations
    # (e.g., from type instabilities)
    let
        t = sol.t[end]
        u_ode = sol.u[end]
        du_ode = similar(u_ode)
        @test (@allocated Trixi.rhs!(du_ode, u_ode, semi, t)) < 1000
    end
  end

  @trixi_testset "elixir_euler_blast_wave.jl with DissipationMatrixWintersEtal" begin
    @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_blast_wave.jl"),
                        surface_flux=FluxPlusDissipation(flux_ranocha,
                                                         DissipationMatrixWintersEtal()),
                        cfl=0.5, tspan=(0.0, 0.01),
                        l2=[
                            0.0240344741630761,
                            0.026530873906038414,
                            0.026530644332116007,
                            0.08614440613707544
                        ],
                        linf=[
                            1.9656605818508415,
                            1.217300641837901,
                            1.217300641837901,
                            2.9173277369453516
                        ])

```

ranocha reviewed (changes requested):
Thanks! Close to be ready to be merged

andrewwinters5000 reviewed (commented):
LGTM once the changes suggested by Hendrik are addressed. I also implemented a local version using `orientation::Integer` and got the same blast wave shape with the pretty significant carbuncles
